### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.reduce_sum`/`T.reduce_max` with `batch > 1` silently returns a wrong reduction instead of the correct one when the reduce dimension spans fewer threads than the block

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2524]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

- Python 3.12, linux
- GPU: NVIDIA H100 (compute capability 9.0, `sm_90`)

```python
import sys, tilelang, torch
print(sys.version.split()[0], sys.platform)   # 3.12.3 linux
print(tilelang.__version__)                   # 0.1.11
print(torch.__version__)                       # 2.12.1+cu130, CUDA 13.0
```

### Problem description

The batched `AllReduce` lowering sizes and strides its shared-memory workspace by `reducing_threads`, while the device template indexes that workspace by `threadIdx.x` over the **full block**. So whenever the reduce dimension spans fewer threads than the block (`reducing_threads < total threads`), a `batch > 1` reduction (`T.reduce_sum`, `T.reduce_max`, …) **silently returns a wrong result** — the per-batch slices overlap each other and run off the end of the workspace. `batch = 1` is correct on the exact same kernel.

The error is a gross value error, not floating-point rounding (the repro uses integer-valued inputs, for which any reassociation is still exact), and it is fully deterministic (bit-identical across repeated launches).

<details>
<summary>Trigger-boundary table (3D row-reduce, <code>dim=1</code>, fp32, integer inputs)</summary>

| config | emitted `AllReduce` | `reducing_threads` / block | `batch=1` | `batch=2` | verdict (batch=2) |
|--------|---------------------|----------------------------|-----------|-----------|-------------------|
| `R=8  W=16 T=128` | `<SumOp, 64, 8, 0, …, 2, 64>::run_batch` | 64 / 128 | ndiff 0 | ndiff 61/256, maxabs 16 | **WRONG** |
| `R=16 W=16 T=128` | `<SumOp, 64, 4, 0, …, 2, 64>::run_batch` | 64 / 128 | ndiff 0 | ndiff 62/256, maxabs 21 | **WRONG** |
| `R=8  W=32 T=128` | `<SumOp, 64, 8, 0, …, 2, 64>::run_batch` | 64 / 128 | ndiff 0 | ndiff 123/512, maxabs 15 | **WRONG** |
| `R=32 W=16 T=128` | `<SumOp, 128, 4, 0, …, 2, 128>::run_batch` | 128 / 128 | ndiff 0 | ndiff 0 | correct (`reducing_threads == total`, no aliasing) |
| `R=8  W=16 T=64`  | `<SumOp, 32, 4, 0, …, 2, 32>::run_batch` | 32 / 64 | ndiff 0 | ndiff 0 | correct (`reducing_threads ≤ 32` → warp-shuffle, no `red_buf`) |

The defect fires **iff** the batched path is taken (`batch > 1`) **and** `32 < reducing_threads < total threads` (so the shared-memory butterfly with `red_buf` is used and `workspace_stride < block`). It is correct when `reducing_threads ≤ 32` (warp-shuffle path, no `red_buf`) or `reducing_threads == total threads` (stride equals the block, no aliasing). `T.reduce_max` reproduces identically. Row `R=8 W=16 T=128` above is the pasted repro; the other rows are inferred from the same lowering with the parameters swapped.

</details>

### Root cause

When batching kicks in, the reduction is lowered to `tl::AllReduce<Reducer, reducing_threads, scale, 0, Barrier, batch, workspace_stride>::run_batch(...)`. The lowering [passes `workspace_stride = reducing_threads`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L603-L605) (the 7th argument to `MakeBatchAllReduce`) and [sizes the workspace `ws_size = reducing_threads * eff_batch`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L611-L614). That SoA layout ("slot for batch item b, thread t = `red_buf[b * reducing_threads + t]`") only holds if `t` is the *reduce-local* thread index `0 .. reducing_threads-1`. The device template instead uses the **block-global** index: [`butterfly_reduce_batch`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/tl_templates/cuda/reduce.h#L268-L295) stores `red_buf[(threadIdx.x - thread_offset) + i * workspace_stride] = x[i]` and loads it back at the XOR partner, with `threadIdx.x` ranging over the whole block (`thread_offset = 0`). The template's own doc says [`workspace_stride` should be *"typically total threads in the block"*](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/tl_templates/cuda/reduce.h#L217-L222), not `reducing_threads`.

The scalar `run` path is immune precisely because it sizes the workspace by the full block: it [allocates `workspace_size = T.thread_bounds->extent`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L791-L796) and the [scalar template stores a single value at `red_buf[threadIdx.x - thread_offset]`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/tl_templates/cuda/reduce.h#L250-L266), which fits exactly with no aliasing — which is why `batch = 1` is correct and only `batch > 1` is wrong.

<details>
<summary>Concrete index walk: <code>reducing_threads = 64</code> in a 128-thread block</summary>

- Workspace allocated: `reducing_threads * eff_batch = 64 * 2 = 128` floats.
- `workspace_stride = 64`, but `threadIdx.x` runs `0 .. 127`.
- The `i = 1` store range is `[threadIdx.x + 64]` for `threadIdx.x ∈ [0, 127]` → indices `[64, 191]`, which (a) **runs off** the 128-element workspace for `threadIdx.x ≥ 64`, and (b) **aliases the `i = 0` slots** of the upper-half threads — e.g. thread 0's `i = 1` slot (index 64) collides with thread 64's `i = 0` slot (index 64).

</details>

### Reproducible example code

A 3D row-reduce over `dim=1` where the reduce dimension maps to 64 threads inside a 128-thread block, so the batched `AllReduce` uses the shared-memory butterfly with `workspace_stride = 64 < 128`. Inputs are integers in `[0, 8)` so a reassociated sum is still exact — any difference is a real miscompile, not float non-associativity. Only `batch` changes between the correct control (`BATCH=1`) and the wrong run (`BATCH=2`).

```python
import torch
import tilelang
import tilelang.language as T


def make_kernel(BATCH):
    @T.prim_func
    def main(A: T.Tensor((16, 8, 16), "float32"), C: T.Tensor((16, 16), "float32")):
        with T.Kernel(8, threads=128) as bx:
            a_s = T.alloc_shared((2, 8, 16), "float32")
            a_f = T.alloc_fragment((2, 8, 16), "float32")
            c_f = T.alloc_fragment((2, 16), "float32")
            T.copy(A[bx * 2, 0, 0], a_s)
            T.copy(a_s, a_f)
            T.reduce_sum(a_f, c_f, dim=1, batch=BATCH)   # flip BATCH 1 -> 2
            T.copy(c_f, C[bx * 2, 0])
    return main


torch.manual_seed(0)
A = torch.randint(0, 8, (16, 8, 16)).float().cuda()   # exact integers
ref = A.sum(dim=1).cpu()                               # (16, 16)

for BATCH in (1, 2):
    kernel = tilelang.compile(make_kernel(BATCH), target="cuda")
    C = torch.empty((16, 16), dtype=torch.float32, device="cuda")
    kernel(A, C)
    diff = (C.cpu() - ref).abs()
    ndiff = int((diff > 1e-4).sum())
    print(f"BATCH={BATCH}: ndiff={ndiff}/{ref.numel()}  maxabs={diff.max().item():g}")

# -> BATCH=1: ndiff=0/256   maxabs=0     (control: scalar path, correct)
# -> BATCH=2: ndiff=61/256  maxabs=16    (batched path, silently wrong)
```

The corrupted rows are exactly the second element of each 2-row tile (output rows 1, 3, 5, …, 15) — the `i = 1` batch element whose workspace slots are aliased; the `i = 0` element (even rows) survives. The result is bit-identical across repeated launches, so this is a deterministic codegen bug, not a synchronization race.

<details>
<summary>Example wrong row (output row 1, <code>BATCH=2</code>)</summary>

For `BATCH=2`, output row 1 is `[27, 20, 23, 46, 29, 27, 16, 14, 29, 20, 21, 34, 25, 33, 19, 21]` but the correct `A[1].sum(dim=0)` is `[20, 20, 25, 46, 22, 27, 19, 14, 19, 20, 15, 34, 29, 33, 23, 21]` — many entries off by 2–10 (e.g. col 0: 27 vs 20).

</details>

<details>
<summary>The two kernels differ only in the emitted reduction call</summary>

```cuda
// BATCH=1 (correct): scalar path, workspace sized to the full block
c_f[i] = tl::AllReduce<tl::SumOp, 64, 8, 0, tl::NamedBarrier<128>>::run(c_f[i], (&(((float*)workspace)[0])));

// BATCH=2 (wrong): batched path, workspace_stride = reducing_threads (64) < block (128)
tl::AllReduce<tl::SumOp, 64, 8, 0, tl::NamedBarrier<128>, 2, 64>::run_batch((&(c_f[0])), (&(((float*)workspace)[0])));
```

</details>

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

A `batch > 1` reduction should produce the same correct result as `batch = 1` on the same kernel — or be rejected at compile time — rather than silently aliasing batch slices in the workspace. The scalar `run` path is already correct here for every config above because it sizes and strides its shared workspace by the full block thread extent ([`workspace_size = T.thread_bounds->extent`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L791-L796)); the batched path should do the same so that the per-thread store index `threadIdx.x - thread_offset` always lands inside its own batch slice.

### Additional context

**The essence: an allocation/indexing coordinate-system mismatch.** The batched workspace is *allocated* in one coordinate system but *indexed* in another:

- **Allocated** by the lowering using the *reduce-local* count `reducing_threads` — both stride and size are `reducing_threads` (`= 64` in the repro). This is the natural, **non-redundant** choice: only `reducing_threads` values per batch item logically need a shared-memory slot, so `reducing_threads × batch` slots is exactly enough.
- **Indexed** by the device template using the *block-global* thread id `threadIdx.x - thread_offset`, which ranges over the **whole block** (`0 .. blockDim-1` = `0..127`) with no fold back into `0 .. reducing_threads-1`. The store `red_buf[(threadIdx.x - thread_offset) + i*stride] = x[i]` is unguarded, so every thread in the block executes it and the indices actually written span the full block, not just `reducing_threads`.

The two agree only when `reducing_threads == blockDim` — which is exactly why the bug hides at that boundary, and why it never appears for `reducing_threads ≤ 32` (the warp-shuffle path uses no workspace at all). Whenever `reducing_threads < blockDim`, the block-global indices overflow the `reducing_threads`-wide slice and collide with the next batch item's slice.

**In principle** the compact allocation is the *right* design — it is not redundant; `reducing_threads × batch` slots are all that is logically needed. It only requires the two sides to share the reduce-local coordinate system. **In practice** the store side never folds `threadIdx.x` into a reduce-local index, so the compact allocation is under-sized for the indices it actually receives. Either side can be brought into agreement (the two options below): raise the allocation to match the block-global indexing, or fold the indexing down to match the compact allocation.

**Suggested fix (proposed, not built).** The simplest fix that matches the device template's `threadIdx.x`-based indexing is to make `workspace_stride` (and the workspace size) the full block thread extent rather than `reducing_threads`, mirroring the scalar path. The two edits land in `src/backend/common/op/reduce.h` — the 7th argument of the [`MakeBatchAllReduce` call at L603-L605](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L603-L605) and the [`ws_size` at L611-L614](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/backend/common/op/reduce.h#L611-L614):

```diff
   std::string allreduce = Impl::MakeBatchAllReduce(
       reducer, reducing_threads, *scale, thread_offset,
-      T.thread_bounds->extent, eff_batch, reducing_threads, T.target);
+      T.thread_bounds->extent, eff_batch,
+      static_cast<int>(*as_const_int(T.thread_bounds->extent)), T.target);
   ...
   if (need_workspace) {
-    int ws_size = reducing_threads * eff_batch;
+    int ws_size =
+        static_cast<int>(*as_const_int(T.thread_bounds->extent)) * eff_batch;
     workspace = T.AddWorkspace(ws_size, ws_dtype);
   }
```

Alternatively, keep the compact `reducing_threads` stride but have the device template ([`butterfly_reduce_batch`](https://github.com/tile-ai/tilelang/blob/a4399e4c9934edd4e34bde89628eba9377548b10/src/tl_templates/cuda/reduce.h#L268-L295)) fold `threadIdx.x` down to `0 .. reducing_threads-1` before the `+ i * workspace_stride` offset, or restrict the batched path to `reducing_threads == total threads`. None of these has been built or verified end-to-end.

**Provenance.** The batched `AllReduce` path was introduced by [#1976](https://github.com/tile-ai/tilelang/pull/1976) ("[Feature] Batched AllReduce for better T.reduce performance", merged 2026-04-28), which chose `workspace_stride = reducing_threads` in the lowering while the same PR's device template indexes by the block-global `threadIdx.x - thread_offset`; the two assumptions only agree when `reducing_threads == total threads`. The line is unchanged at commit `a4399e4c` (2026-06-10) and reproduces on the released `0.1.11` wheel.

**Dedup.** I searched the open and closed tracker and found no existing report of this `batch`-driven workspace-aliasing defect. Distinct from **#2294** (*"Strided reduce"*, closed) — that is the **bf16x2 vectorized-packing** path's `packed_offset` index-remapping error (triggered by an odd "rounds" factor with `num_vectorize > 1`); this bug is the **batched-workspace** path's `workspace_stride` using `reducing_threads` instead of the block extent (triggered by `batch > 1` with `32 < reducing_threads < block`, and dtype-independent — it reproduces with integer inputs). Different sub-path, variable, and trigger within `AllReduce`.
